File: backend/common/src/main/java/com/bankflow/common/iban/Iban.java
package com.bankflow.common.iban;

import java.util.Map;

/**
 * Checks that an IBAN could exist before a payment is built around it.
 *
 * An IBAN is self-checking by design: two of its characters are a checksum
 * over the rest, so a mistyped digit is detectable without asking anybody.
 * That is the entire reason the format exists, and not using it means a
 * payment leaves on an account number that was never real.
 *
 * The scheme will not catch this for us. ISO 20022 types IBAN as
 * {@code [A-Z]{2}[0-9]{2}[a-zA-Z0-9]{1,30}}, which any plausible-looking
 * string satisfies -- so a wrong-length Kosovo IBAN passes XSD validation,
 * goes to KIPS, and comes back days later as a rejection the customer has
 * been waiting on. Validating here turns that into an error at the keyboard.
 */
public final class Iban {

    /**
     * Length of a complete IBAN per country, from the SWIFT IBAN registry.
     *
     * Not every country in the world -- the ones this bank can actually reach.
     * An unknown country is not rejected outright, because being absent from
     * this list says something about the list, not the IBAN; it still has to
     * be structurally sound and pass its checksum.
     */
    private static final Map<String, Integer> LENGTHS = Map.ofEntries(
            Map.entry("XK", 20),
            Map.entry("AL", 28),
            Map.entry("AT", 20),
            Map.entry("BA", 20),
            Map.entry("BE", 16),
            Map.entry("BG", 22),
            Map.entry("CH", 21),
            Map.entry("CY", 28),
            Map.entry("CZ", 24),
            Map.entry("DE", 22),
            Map.entry("DK", 18),
            Map.entry("EE", 20),
            Map.entry("ES", 24),
            Map.entry("FI", 18),
            Map.entry("FR", 27),
            Map.entry("GB", 22),
            Map.entry("GR", 27),
            Map.entry("HR", 21),
            Map.entry("HU", 28),
            Map.entry("IE", 22),
            Map.entry("IT", 27),
            Map.entry("LT", 20),
            Map.entry("LU", 20),
            Map.entry("LV", 21),
            Map.entry("ME", 22),
            Map.entry("MK", 19),
            Map.entry("MT", 31),
            Map.entry("NL", 18),
            Map.entry("NO", 15),
            Map.entry("PL", 28),
            Map.entry("PT", 25),
            Map.entry("RO", 24),
            Map.entry("RS", 22),
            Map.entry("SE", 24),
            Map.entry("SI", 19),
            Map.entry("SK", 24),
            Map.entry("SM", 27),
            Map.entry("TR", 26)
    );

    /** The widest an IBAN may be under ISO 13616, and the narrowest in use. */
    private static final int ABSOLUTE_MIN = 15;
    private static final int ABSOLUTE_MAX = 34;

    private Iban() {
    }

    /**
     * Spaces are how people write IBANs and not how they are stored.
     *
     * Printed in groups of four on every bank statement, so a pasted one
     * arrives with spaces in it. Stripping them is kinder than refusing.
     */
    public static String normalise(String iban) {

        if (iban == null) {
            return null;
        }

        return iban.replace(" ", "").replace(" ", "").trim().toUpperCase();
    }

    /**
     * Why this is not an IBAN, or null if it is one.
     *
     * Returns the reason rather than a boolean because every one of these is
     * something the person typing can act on, and "invalid IBAN" is not.
     */
    public static String reasonInvalid(String raw) {

        String iban = normalise(raw);

        if (iban == null || iban.isEmpty()) {
            return "An IBAN is required";
        }

        if (!iban.matches("[A-Z]{2}[0-9]{2}[A-Z0-9]+")) {
            return "An IBAN starts with two letters for the country and two "
                    + "check digits, then the account number";
        }

        if (iban.length() < ABSOLUTE_MIN || iban.length() > ABSOLUTE_MAX) {
            return "An IBAN is between %d and %d characters; this one is %d"
                    .formatted(ABSOLUTE_MIN, ABSOLUTE_MAX, iban.length());
        }

        String country = iban.substring(0, 2);
        Integer expected = LENGTHS.get(country);

        if (expected != null && iban.length() != expected) {
            return "A %s IBAN is %d characters; this one is %d".formatted(
                    country, expected, iban.length()
            );
        }

        if (mod97(iban) != 1) {
            return "The check digits do not match the rest of the IBAN, so "
                    + "something in it has been mistyped";
        }

        return null;
    }

    public static boolean isValid(String raw) {
        return reasonInvalid(raw) == null;
    }

    /** Whether this IBAN belongs to the given country. */
    public static boolean isFromCountry(String raw, String country) {

        String iban = normalise(raw);

        return iban != null
                && iban.length() >= 2
                && iban.startsWith(country);
    }

    /**
     * ISO 7064 MOD-97-10, which a valid IBAN leaves at 1.
     *
     * The first four characters move to the end and letters become numbers
     * from A=10, giving a number far too big for a long -- so the remainder is
     * carried digit by digit instead.
     */
    private static int mod97(String iban) {

        String rearranged = iban.substring(4) + iban.substring(0, 4);

        int remainder = 0;

        for (int i = 0; i < rearranged.length(); i++) {

            char character = rearranged.charAt(i);

            if (character >= '0' && character <= '9') {
                remainder = (remainder * 10 + (character - '0')) % 97;
            } else {
                int value = character - 'A' + 10;
                remainder = (remainder * 100 + value) % 97;
            }
        }

        return remainder;
    }
}

File: frontend/src/app/core/validation/iban.ts
/**
 * The same IBAN and scheme-text rules the server enforces.
 *
 * Duplicated deliberately rather than asked for over the wire: the point is to
 * say so while the field still has the cursor in it. The server remains the
 * authority and repeats every check — this only saves the round trip needed to
 * be told what the form already knows.
 */

/** Complete IBAN length per country, from the SWIFT IBAN registry. */
export const IBAN_LENGTHS: Record<string, number> = {
  XK: 20, AL: 28, AT: 20, BA: 20, BE: 16, BG: 22, CH: 21, CY: 28, CZ: 24,
  DE: 22, DK: 18, EE: 20, ES: 24, FI: 18, FR: 27, GB: 22, GR: 27, HR: 21,
  HU: 28, IE: 22, IT: 27, LT: 20, LU: 20, LV: 21, ME: 22, MK: 19, MT: 31,
  NL: 18, NO: 15, PL: 28, PT: 25, RO: 24, RS: 22, SE: 24, SI: 19, SK: 24,
  SM: 27, TR: 26,
};

/** IBANs are printed in groups of four, so pasted ones arrive with spaces. */
export function normaliseIban(value: string | null | undefined): string {
  return (value ?? '').replace(/[\s ]/g, '').toUpperCase();
}

/**
 * ISO 7064 MOD-97-10. A valid IBAN leaves a remainder of 1.
 *
 * The number is far too large for a JS number, so the remainder is carried
 * along digit by digit instead of being computed in one go.
 */
function mod97(iban: string): number {

  const rearranged = iban.slice(4) + iban.slice(0, 4);

  let remainder = 0;

  for (const character of rearranged) {

    if (character >= '0' && character <= '9') {
      remainder = (remainder * 10 + (character.charCodeAt(0) - 48)) % 97;
    } else {
      remainder = (remainder * 100 + (character.charCodeAt(0) - 55)) % 97;
    }
  }

  return remainder;
}

/** Why this is not an IBAN, or null if it is one. */
export function ibanProblem(value: string | null | undefined): string | null {

  const iban = normaliseIban(value);

  if (!iban) {
    return null;
  }

  if (!/^[A-Z]{2}[0-9]{2}[A-Z0-9]+$/.test(iban)) {
    return 'An IBAN starts with two letters for the country and two check '
      + 'digits, then the account number.';
  }

  if (iban.length < 15 || iban.length > 34) {
    return `An IBAN is between 15 and 34 characters; this one is ${iban.length}.`;
  }

  const country = iban.slice(0, 2);
  const expected = IBAN_LENGTHS[country];

  if (expected && iban.length !== expected) {
    return `A ${country} IBAN is ${expected} characters; this one is ${iban.length}.`;
  }

  if (mod97(iban) !== 1) {
    return 'The check digits do not match the rest of the IBAN, so something '
      + 'in it has been mistyped.';
  }

  return null;
}

/** KIPS clears Kosovo only, so a foreign IBAN cannot go on a domestic rail. */
export function wrongCountryForRail(
  value: string | null | undefined,
  rail: string,
): string | null {

  const iban = normaliseIban(value);

  if (!iban || iban.length < 2) {
    return null;
  }

  const domestic = rail === 'KIPS_ACH' || rail === 'KIPS_RTGS';

  if (domestic && !iban.startsWith('XK')) {
    return 'KIPS clears Kosovo payments only. Send this as an international '
      + 'payment instead.';
  }

  return null;
}

const BIC = /^[A-Z]{4}[A-Z]{2}[A-Z0-9]{2}([A-Z0-9]{3})?$/;

export function bicProblem(value: string | null | undefined): string | null {

  const bic = (value ?? '').trim().toUpperCase();

  if (!bic) {
    return null;
  }

  return BIC.test(bic)
    ? null
    : 'A BIC is 8 or 11 characters: four for the bank, two for the country, '
      + 'two for the location, and optionally three for a branch.';
}

/**
 * The characters ISO 20022 restricted text will carry.
 *
 * Not folded silently here the way an outgoing statement folds them: changing
 * the beneficiary's name changes who is being paid, so it is refused and said
 * out loud instead.
 */
const PERMITTED = /^[0-9a-zA-Z/\-?:().,'+ ]*$/;

export function schemeTextProblem(
  value: string | null | undefined,
  label: string,
  max: number,
): string | null {

  const text = value ?? '';

  if (!text) {
    return null;
  }

  if (text.length > max) {
    return `${label} must be ${max} characters or fewer; this is ${text.length}.`;
  }

  if (!PERMITTED.test(text)) {

    const offending = [...new Set(
      [...text].filter(character => !PERMITTED.test(character)),
    )].join('');

    return `${label} contains characters the scheme will not carry: ${offending}. `
      + `Letters a-z, digits, and / - ? : ( ) . , ' + are allowed.`;
  }

  return null;
}

/** Money has two decimal places; the ledger stores no more than that. */
export function amountProblem(value: number | null): string | null {

  if (value === null || value === undefined) {
    return null;
  }

  if (value <= 0) {
    return 'The amount must be greater than zero.';
  }

  const decimals = (String(value).split('.')[1] ?? '').length;

  return decimals > 2
    ? 'The amount cannot be more precise than a cent.'
    : null;
}
